### DISCUSSION TITLE
Mixing hardware & software i2c on same port pins

### DISCUSSION THREAD

**Discussion by ADenislebel** [micropython/micropython#13730] [Hardware & Peripherals]
Hello, my platform is RP2.  I can access 4 x  HTU21 sensors via a pca9548 multiplex. I got that working first using code obtained on GitHub.  The driver code uses the hardware based I2C functions to access both the mux and subsequently the HTUs on the mux output pins. I also need to access 8 output IOs via a PCF8574  expander on the same bus (same SCL & SCK pins). Needless to say the Mux and the expander have different I2C addresses. Both I2C access work separately. However when I attempt access both I2C ports, the first access works but the other access fails. Whichever I run first cause the other to fail. When the expander is accessed second I get a writeto() error on the first attempt to write to the expander. When the Mux+Htu is accessed second it give a timeout error on writing to the mux when setting the chnl/snsr that I want to read.  My question is it because the Hardware and Software I2C drivers can't co-exist on the same port. And if so is there a work around possible. At the moment I'm considering having to re-initialize each I2C port on every write or read. Will that work or is there other issue to consider.
P.S. I did not include code example because I'm on mission to understand and learn.
Thanks Denis Lebel

**Comment by robert-hh** [comment]
That sounds strange. MUX and Expander should just be two different devices on the I2C bus. You just have to create the I2C object only once and use it for both devices. In any case, you cannot have both hardware and software I2C on the same pins.

**Reply by ADenislebel** [reply]
Well you just confirm what suspected. Indeed I did attempt to use the same I2C for both jobs. However when I attempt to write a  single byte to the expander (to set the HTU channel) I got buffer protocol errors using the hardware based I2C.write() function. I was able to get around it by using the software SoftI2C.writeto(). I will have to do my homework to make it work with the hardware I2C.
Thanks...Denis

**Comment by ADenislebel** [comment]
I got it resolved by doing exactly that. I created a short py to just
create the I2C object and imported into both pys that require access to the
i2c bus. Thanks again and have a nice day...Denis
